LCM/scripts/InstallModule.py

```diff
@@ -16,188 +25,293 @@ def usage():
+            exitWithError("Unable to find the resource library folder for the resource " + resource + " and platform architecture " + resourceArchitectureFolderName + " at the path " + resourceScriptsFolderPath + " in the extracted module.")
+
+        resourceScriptsFolderPath = join(resourceLibraryFolderPath, 'Scripts')
+        if not os.path.isdir(resourceLibraryFolderPath):
+            exitWithError("Unable to find the resource library scripts folder for the resource " + resource + " and platform architecture " + resourceArchitectureFolderName + " at the path " + resourceScriptsFolderPath + " in the extracted module.")
+
+        pythonVersionFileNames = ['2.4x-2.5x', '2.6x-2.7x', '3.x']
+
+        for pythonVersionFileName in pythonVersionFileNames:
+            resourceScriptsPythonVersionFolderPath = join(resourceScriptsFolderPath, pythonVersionFileName)
+            if not os.path.isdir(resourceScriptsPythonVersionFolderPath):
+                exitWithError("Unable to find the version-specific Python folder under the resource library scripts folder for the resource " + resource + " and platform architecture " + resourceArchitectureFolderName + " at the path " + resourceScriptsPythonVersionFolderPath + " in the extracted module.")
+
+            resourceScriptsPythonVersionScriptsFolderPath = join(resourceScriptsPythonVersionFolderPath, 'Scripts')
+            if not os.path.isdir(resourceScriptsPythonVersionScriptsFolderPath):
+                exitWithError("Unable to find the version-specific Python scripts folder under the resource library scripts folder for the resource " + resource + " and platform architecture " + resourceArchitectureFolderName + " at the path " + resourceScriptsPythonVersionScriptsFolderPath + " in the extracted module.")
+
+            resourceScriptFileNames = os.listdir(resourceScriptsPythonVersionScriptsFolderPath)
+
+            dscScriptsPythonVersionFolderPath = join(helperlib.DSC_SCRIPT_PATH, pythonVersionFileName)
+            dscScriptsPythonVersionScriptsFolderPath = join(dscScriptsPythonVersionFolderPath, 'Scripts')
+
+            for resourceScriptFileName in resourceScriptFileNames:
+                resourceScriptFilePath = join(resourceScriptsPythonVersionScriptsFolderPath, resourceScriptFileName)
+                if (os.path.isfile(resourceScriptFilePath)):
+                    shutil.copy(resourceScriptFilePath, dscScriptsPythonVersionScriptsFolderPath)
+
+        # Copy the resource library file to the OMI library folder
+        resourceLibraryFileName = "lib" + resource + ".so"
+        resourceLibraryFileSourcePath = join(resourceLibraryFolderPath, resourceLibraryFileName)
+
+        if helperlib.DSC_NAMESPACE == "root/Microsoft/DesiredStateConfiguration":
+            resourceLibraryFileDestinationName = resourceLibraryFileName
+        else:
+            resourceLibraryFileDestinationName = resourceLibraryFileName.replace('.so', "_" + omiNamespaceFolderName + ".so")
+
```

Comment:
we are renaming the binary during copy instead of generating it during make process with new names. the library so metadata will probably has it old name and this can cause problem when doing debugging etc with these binaries. do we know why we fixed it this way instead of generating it through build process ? can we test if debugging works  with these libraries ?

Reply:
i think this was one of the open question during spec review ?

---
In reply to: [174943306](https://github.com/Microsoft/PowerShell-DSC-for-Linux/pull/366#discussion_r174943306) [](ancestors = 174943306)

Reply:
Ultimately I decided to put it in the installer instead because changing the library name only at build time would require all of our current partners who publish resource modules that get pulled to re-build all their modules. The SxS issue would then not be fixed for any of the pulled modules until new versions of all those modules were published and available to the public.

With the fix in the installer, none of our partner teams will need to publish new modules. The current ones will work and the fix will happen for all modules at once.

Reply:
There is another way of addressing what you said.
1- build the binaries with new names
2- for oms scenario where you are renaming the binaries ... check binary name if it already conform to new naming convention do not touch it. If it does not then rename it just as a workaround. 
3- In parallel work with partners to update the modules for pull. during transition period when all modules with new names are not published (2) above will be the workaround. once everyone is updated we can remove above workaround from the InstallModule script.

Reply:
This script fix ensures that no matter what the customer names their resource (like say they named it resource_root-oms so the lib happens to be resource_root-oms.so for DIY) then STILL the resource will work in both DIY DSC and OMS DSC with no conflict. If I added in a hack to say 'if it already has this name then don't rename it' then that resource would break because it would not get an extra '_root-oms' tacked on to the end of the OMS library name.

Then even if we suddenly had some 3rd version of Linux DSC that someone was working on or if we need multiple instances of DSC running on the same box under different workspaces with different resource versions, the fix in this script will work without re-building for every DSC namespace. You can just install modules into the new side-by-side DSC without having to re-build the resources.